src/Metadata/Property/Factory/CachedPropertyNameCollectionFactory.php

```diff
@@ -0,0 +1,66 @@
+<?php
+
+/*
+ * This file is part of the API Platform project.
+ *
+ * (c) Kévin Dunglas <dunglas@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace ApiPlatform\Core\Metadata\Property\Factory;
+
+use ApiPlatform\Core\Metadata\Property\PropertyNameCollection;
+use Psr\Cache\CacheException;
+use Psr\Cache\CacheItemPoolInterface;
+
+/**
+ * Caches property name collection.
+ *
+ * @author Teoh Han Hui <teohhanhui@gmail.com>
+ */
+final class CachedPropertyNameCollectionFactory implements PropertyNameCollectionFactoryInterface
+{
+    const CACHE_KEY_PREFIX = 'property_name_collection_';
+
+    private $cacheItemPool;
+    private $decorated;
+
+    public function __construct(CacheItemPoolInterface $cacheItemPool, PropertyNameCollectionFactoryInterface $decorated)
+    {
+        $this->cacheItemPool = $cacheItemPool;
+        $this->decorated = $decorated;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function create(string $resourceClass, array $options = []) : PropertyNameCollection
+    {
+        $cacheKey = self::CACHE_KEY_PREFIX.md5(json_encode([$resourceClass, $options]));
```

Comment:
Switched to `sha1` since it's (surprisingly) faster than `md5` (at least that's what `openssl speed md5 sha1` shows on my machine and a Digital Ocean droplet).